### PR TITLE
ENG-612: Token Studio backend support: return details from synchronizeTokensFromData

### PR DESCRIPTION
**Pull request by tchief**
### Changes
- Added `synchronizeTokensFromDataWithResults` method that returns imported results data from `mergeWithRemoteSource`.
  - Existing `synchronizeTokensFromData` should not be changed.
  - Keep `stop on first failure` behavior.
- Added `SupernovaToolsDesignTokensResult` as container for result data
- Export few related types
- Bump version

### CODE REVIEW

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts [hunk 1]
```diff
@@ -34,6 +34,13 @@ export type SupernovaToolsDesignTokensLoadingResult = {
   groups: Array<TokenGroup>
 }
 
+export type SupernovaToolsDesignTokensResult = {
```
Comment (tchief) [inline src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:37]: Another option is to not return tokens, but use slimmer type, like this
```
export type SupernovaToolsDesignTokensResultSlim = {
    map: DTPluginToSupernovaMap
    tokensCreated: number;
    tokensUpdated: number;
    tokensDeleted: number;
  }
```
and fill it like this
```
  results.push({
    map, 
    tokensCreated: mergeResult.diff.toCreate.length,
    tokensUpdated: mergeResult.diff.toUpdate.length,
    tokensDeleted: mergeResult.diff.toDelete.length
  });
  // ...
  results.push({
    map, 
    tokensCreated: 0,
    tokensUpdated: mergeResult.tokens.length,
    tokensDeleted: 0
  });
```

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts [hunk 2]
```diff
@@ -34,6 +34,13 @@ export type SupernovaToolsDesignTokensLoadingResult = {
...
+  map: DTPluginToSupernovaMap
```
Comment (tchief) [inline src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:38]: Another option to remove nodes properties from `DTPluginToSupernovaMap`
```
Omit<DTPluginToSupernovaMap, "nodes" | "processedNodes" | "processedGroups">
```

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts [hunk 3]
```diff
@@ -404,7 +427,8 @@ export class SupernovaToolsDesignTokensPlugin {
     }
 
     return {
-      theme: theme
+      theme: theme,
+      tokens: themeMergerResult.overriddenTokens
```
Comment (tchief) [inline src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:431]: Would `theme.overridenTokens` contain same data as `themeMergerResult.overriddenTokens`? If so, I can remove this new property and get everything from `theme`.

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts [hunk 4]
```diff
@@ -84,6 +91,17 @@ export class SupernovaToolsDesignTokensPlugin {
     mapping: DTPluginToSupernovaMapPack,
     settings: DTPluginToSupernovaSettings
   ): Promise<boolean> {
+    await this.synchronizeTokensFromDataWithResults(data, mapping, settings)
+    return true
+  }
+
+  async synchronizeTokensFromDataWithResults(
```
Comment (tchief) [inline src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:98]: Created separate method, not sure whether this one is used by external clients. Could leave just one method if that's better.

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts (hunk 2 above)
Comment (tchief) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:38]: We also might send less data in this case (exclude `nodes`)

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts (hunk 1 above)
Comment (JiriTrecak) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:37]: I think we should do a middle ground - use your slim maps, but instead of `number`, use tokens.map(t => t.id) from SDK. This way we can also reference tokens that were created/deleted etc. - WDYT?

I think there is value in keeping the reference to what was what.

Comment (JiriTrecak) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:37]: + of course don't call it slim :)

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts (hunk 2 above)
Comment (JiriTrecak) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:38]: I agree, raw data are not at all useful for the result

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts (hunk 4 above)
Comment (JiriTrecak) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:98]: @tchief I would only keep the method that handles the data, and completely remove methods that handle loading from file and loading from directory. The reason being that I want to remove `fs` package from the dependencies so SDK can be used in the browser - then, we can move the `fs` handling to the CLI to pre-process the data for the user and use the same routines that were used here (I'll do that part). For consumer, nothing will change because they are not using SDK directly, only through CLI.

So basically, here:
https://github.com/Supernova-Studio/cli/blob/main/src/commands/sync-tokens.ts#L86
We'd change to load from data, and move the functions to pre-process the data to the CLI.

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts (hunk 3 above)
Comment (JiriTrecak) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:431]: I'd keep the merger result, because it might not be guaranteed going forward this will be the case.

**Review by JiriTrecak — CHANGES_REQUESTED** [review]
Left few comments, thank you!

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts (hunk 4 above)
Comment (tchief) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:98]: @JiriTrecak Maybe we should do this in separate PR, so we can release new SDK so we can deploy API to dev to test, without depending on making CLI changes and releasing it? WDYT?

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts (hunk 1 above)
Comment (tchief) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:37]: Is this better?
```
export type SupernovaToolsDesignTokensResult = {
    map: Pick<DTPluginToSupernovaMap, 'bindToBrand' | 'bindToTheme' | 'pluginSets' | 'pluginTheme' | 'type'>
    tokensCreated: Array<string>;
    tokensUpdated: Array<string>;
    tokensDeleted: Array<string>;
  }
```

We could do `id` for now.
I think we would need more data, than just `id` if we want to somehow process tokens later. At this point it's just a speculation, probably we never need it.

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts [hunk 5]
```diff
@@ -126,15 +150,21 @@ export class SupernovaToolsDesignTokensPlugin {
       if (!theme) {
         throw new Error(`Unknown theme ${map.bindToTheme} provided in binding.\n\nAvailable themes in this design system: ${brands.map(b => `Brand: ${b.name} (id: ${b.persistentId})\n${themes.filter(th => th.brandId == b.persistentId).map(t => `    Theme: ${t.name} (id: ${t.id})`)}`)}`)
       }
-      await this.mergeThemeWithRemoteSource(map.processedNodes, brand, theme, !settings.dryRun, settings.verbose)
+      const mergeResult = await this.mergeThemeWithRemoteSource(map.processedNodes, brand, theme, !settings.dryRun, settings.verbose)
+      results.push({
+        map: _.pick(map, ["bindToBrand", "bindToTheme", "pluginSets", "pluginTheme", "type"]),
+        tokensCreated: [],
+        tokensUpdated: mergeResult.tokens.map(t => t.id),
+        tokensDeleted: []
```
Comment (tchief) [inline src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:158]: @JiriTrecak I consider results of `mergeThemeWithRemoteSource` for theme as `tokensUpdated`
```
console.log(`Created or updated overrides: ${themeMergerResult.overriddenTokens.length}`)
```

**Review by tchief — COMMENTED** [review]
Add comment

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts (hunk 1 above)
Comment (tchief) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:37]: The only concern with just ids: on calling API `/token-studio` all we can return are ids in this case.

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts (hunk 4 above)
Comment (JiriTrecak) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:98]: I agree, let's do it separately

File: src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts (hunk 1 above)
Comment (JiriTrecak) [reply src/tools/design-tokens/SDKToolsDesignTokensPlugin.ts:37]: I think that is fine. Don't expect it to be used nonetheless, not for now, and we can always expand on it.

**Review by JiriTrecak — APPROVED** [review]
